### DISCUSSION TITLE
how adopt resources with werf?

### DISCUSSION THREAD

**Discussion by Fr0zenBoy** [werf/werf#6246] [General]
I'm trying to migrate my applications to werf but we have a lot of things that went through the need to bake the manifest where helm was the templant generator but it wasn't used to apply it so some things like helm release were not present when executing the same template with werf, the labens used to validate the adoption are not present

In older versions of werf like 1.2 there are texts about the 3 forms of merge and several topics about adoption but now in werf 2.0 it seems that this was forgotten or ignored. I didn't find any way to do this other than editing the templates manually or using plugins.

 Error message:
 
 ```
 Error: error processing deployable resources: error validating adoptable resources: adoption validation failed: 6 errors occurred:
	* resource "istio-system/Certificate/app" is not adoptable: annotation "meta.helm.sh/release-name" not found, must be set to "app-dev", annotation "meta.helm.sh/release-namespace" not found, must be set to "default"
 ```

**Comment by ilya-lesikov** [comment]
There is no doc about adoption because I think it does not deserve its own doc, there is not much to tell. All you need to do is explained in the error message. All deployed resources must have two annotations set as described above. Maybe we could even add a flag to forcefully apply new annotations.

In regards to 3-way merge (if that's what you mean) then there is no 3-way merge in werf anymore. We replaced it with Server-Side Apply. But this is just internal mechanism we use to update resources in the cluster. There were several issues with 3-way merge so the user might need to know how it worked to solve these issues, but with Server-Side apply there is none of these, it just works. So I feel there is no need to overburden the user with this. It works as you expect it too - whatever you have in the chart will be applied to the cluster. The only thing that probably worth mentioning is that we reset whatever changes done to the resources with `kubectl edit`, but nothing more.
